# Stage I - Data Understanding and Linking

## What is the U.S. Opioid Epidemic?

- In the late 1990s, pharmaceutical companies reassured the medical community that patients would not become addicted to opioid pain relievers and healthcare providers began to prescribe them at greater rates.

- Increased prescription of opioid medications led to widespread misuse of both prescription and non-prescription opioids before it became clear that these medications could indeed be highly addictive.

![opioid](https://www.hhs.gov/opioids/sites/default/files/inline-images/opioids-infographic.png)

Source: https://www.hhs.gov/opioids/about-the-epidemic/index.html


We are going to study the underlying patterns that exist between opioid related deaths and the different socio-economic, demographic, geographic, and equity related variables that are available for US population. Our goal is to extract such patterns and understand them futher using data science techniques. 

In order to achieve that, the project is separated into 5 stages:

- Stage I - Data and Project Understanding,
- Stage II - Data Modeling,
- Stage III - Distributions and Hypothesis Testing,
- Stage IV - Dashboard




### Opioid overdose dataset linking

In this stage we utilize multiple publicly available datasets and link them together for analytics. Our goal here is to help identify patterns which contribute to drug overdose deaths. Within this notebook we will explore:

1. **Creating Index** - Developing an index key for linking datasets
2. **Join** - Using join to merge data based on index key.

The notebooks is viewable via any browser. 

#### Software used (open source):

- `python` - https://www.python.org/download/releases/3.0/
- `pandas` - https://pandas.pydata.org/
- `plotly` - https://plotly.com/

### Datasets


#### 1. Drug Overdose Dataset

The overdose death/cause dataset was obtained from CDC Wonder (https://wonder.cdc.gov/ucd-icd10.html). The dataset is from the Underlying Cause of Death database contains mortality and population counts for all U.S. counties. Data are based on death certificates for U.S. residents. Each death certificate identifies a single underlying cause of death and demographic data. 

- From this data we obtained the Drug/Alcohol Induced causes data for 2019 across all counties in US. 
- https://wonder.cdc.gov/wonder/help/ucd.html#Drug/Alcohol%20Induced%20Causes
- File: `./data/Underlying Cause of Death-County-2019.txt`

#### 2. County Health Rankings

The County Health Rankings provide a snapshot of a community’s health and a starting point for investigating and discussing ways to improve health. The annual Rankings measures vital health factors, including high school graduation rates, obesity, smoking, unemployment, access to healthy foods, the quality of air and water, income inequality, and teen births in nearly every county in America. The dataset provides a snapshot of how health is influenced by where we live, learn, work and play.

- From this data we obtained the measures data for 2019 across all counties in US. 
- https://www.countyhealthrankings.org/
- Data Dictionary - https://www.countyhealthrankings.org/sites/default/files/DataDictionary_2019.pdf
- File: `./data/County_Health_Ranking.csv`

#### 3. County Opioid Dispensing Rates

The third dataset in this notebook is the Opoid Dispensing Rate dataset. The dataset has geographic distribution of retail opioid prescriptions dispensed per 100 persons per year from 2006–2019. Rates are classified by the Jenks natural breaks classification method into four groups using the 14-year range of data to determine the class breaks. 

- We utilize County Opoid Dispensing Rates for 2019.
- https://www.cdc.gov/drugoverdose/maps/rxcounty2019.html
- File: `./data/2019-Opioid_Rate.csv`

## Tasks:

#### Task 1: (10 pts)
- **T1.1** Initialize a Github Repository for your project. (10 pts)
    - Add a description (readme.MD) to your project. See here on how to setup: https://bulldogjob.com/news/449-how-to-write-a-good-readme-for-your-github-project

#### Task 2: (50 pts)
- Team:
    - **T2.1** Entire team looks at the datasets and understands the type of variables present in each of the data. (10 pts)
        - **Deliverable** 
            - Create a report of what the project is about, why this is an important area of work (in your own words), and how can data science help.
            - Outline how the datasets can be merged together and the common variables. 
        
- Member: 
    - **M2.1** Study prior research in the area. (20 pts)
        - Read https://link.springer.com/article/10.1007/s40265-017-0846-6
        - Select one other paper to study in relation to the area. Use https://scholar.google.com/ to search for papers which are related to the goal of this project.
        - **Deliverable**
            - Prepare a 1 page summary of what was discovered in these two paper. Significant outcomes, i.e. which variables/determinants are linked to opioid endemic. 
    - **M2.2** Each student member of the team selects 10 variables they think that are important from the available dataset. (20 pts) 
        - **Deliverable**
            - Prepare a data dictionary (data and datatype - variable dictionary. https://analystanswers.com/what-is-a-data-dictionary-a-simple-thorough-overview/) of the selected variables
            - Include justification of why you think the variables are important. 

Upload the team and member reports to canvas and your Github Repository. 

#### Task 3: (50 pts)
- Team: (20 pts)
    
    - **T3.1** Create a team notebook to read in the Opioid Mortality data using `pandas` and display the dataframe in a notebook.
    - **T3.2** Normalize the mortality data by population, i.e. number of deaths per 100,000 population. 
    - **T3.3** Identify issues with the data
        - Merge issues, missing values, inconsistent values, etc.
        - Describe solutions to fix it. 
    
- Member: (30 pts)
    - **M3.1** Merge all the three datasets to create a super datafame. (10 pts)
        - Display the super dataframe - Its should be (2527, 542) shape
        - Export it to a csv format.
    - **M3.2** Identify counties and states (top 10) with the highest opioid mortality rates. (20 pts)
        - Use mean and median for counties within states to compare (for the state level).
        - Describe your intution on why the rates are high in these states and counties. 

**Deliverable**
Each member creates separate notebooks for member tasks. Upload all notebooks to Github Repository. 


<b> Importing required packges<b>

In [1]:
import pandas as pd

<b>Reading all datasets<b>
<br> Reading cause_of_death txt file using pandas and giving tab as separator<br>

In [2]:
cause_of_death = pd.read_csv("../../../../data/stage_1/Underlying Cause of Death-County-2019.txt", sep="\t")
cause_of_death

Notes                 County  County Code  \
0       NaN     Autauga County, AL         1001   
1       NaN     Autauga County, AL         1001   
2       NaN     Baldwin County, AL         1003   
3       NaN     Baldwin County, AL         1003   
4       NaN     Baldwin County, AL         1003   
...     ...                    ...          ...   
5536    NaN  Sweetwater County, WY        56037   
5537    NaN       Teton County, WY        56039   
5538    NaN       Uinta County, WY        56041   
5539    NaN       Uinta County, WY        56041   
5540    NaN    Washakie County, WY        56043   

                             Drug/Alcohol Induced Cause  \
0     Drug poisonings (overdose) Unintentional (X40-...   
1          Drug poisonings (overdose) Suicide (X60-X64)   
2     Drug poisonings (overdose) Unintentional (X40-...   
3          Drug poisonings (overdose) Suicide (X60-X64)   
4     Drug poisonings (overdose) Undetermined (Y10-Y14)   
...                                                 ...   
5536       Drug poisonings (overdose) Suicide (X60-X64)   
5537  Drug poisonings (overdose) Unintentional (X40-...   
5538  Drug poisonings (overdose) Unintentional (X40-...   
5539  Drug poisonings (overdose) Undetermined (Y10-Y14)   
5540  Drug poisonings (overdose) Unintentional (X40-...   

     Drug/Alcohol Induced Cause Code  Deaths  Population  Crude Rate  
0                                 D1      69     1087149         6.3  
1                                 D2      14     1087149  Unreliable  
2                                 D1     424     3758097        11.3  
3                                 D2      71     3758097         1.9  
4                                 D4      19     3758097  Unreliable  
...                              ...     ...         ...         ...  
5536                              D2      15      873221  Unreliable  
5537                              D1      13      440125  Unreliable  
5538                              D1      51      426347        12.0  
5539                              D4      10      426347  Unreliable  
5540                              D1      14      172114  Unreliable  

[5541 rows x 8 columns]

In [3]:
health_ranking = pd.read_csv("../../../../data/stage_1/County_Health_Ranking.csv")
health_ranking

State FIPS Code  County FIPS Code  5-digit FIPS Code State Abbreviation  \
0                   1                 0               1000                 AL   
1                   1                 1               1001                 AL   
2                   1                 3               1003                 AL   
3                   1                 5               1005                 AL   
4                   1                 7               1007                 AL   
...               ...               ...                ...                ...   
3188               56                37              56037                 WY   
3189               56                39              56039                 WY   
3190               56                41              56041                 WY   
3191               56                43              56043                 WY   
3192               56                45              56045                 WY   

                   Name  Release Year  County Ranked (Yes=1/No=0)  \
0               Alabama          2019                         NaN   
1        Autauga County          2019                         1.0   
2        Baldwin County          2019                         1.0   
3        Barbour County          2019                         1.0   
4           Bibb County          2019                         1.0   
...                 ...           ...                         ...   
3188  Sweetwater County          2019                         1.0   
3189       Teton County          2019                         1.0   
3190       Uinta County          2019                         1.0   
3191    Washakie County          2019                         1.0   
3192      Weston County          2019                         1.0   

      Premature death raw value  Premature death numerator  \
0                   9917.232898                    80440.0   
1                   8824.057123                      815.0   
2                   7224.632160                     2827.0   
3                   9586.165037                      451.0   
4                  11783.543680                      445.0   
...                         ...                        ...   
3188                7497.439952                      495.0   
3189                3786.128226                      124.0   
3190                7790.302043                      262.0   
3191                5504.650970                      108.0   
3192                4991.168273                       82.0   

      Premature death denominator  ...  Male population 18-44 raw value  \
0                      13636816.0  ...                              NaN   
1                        156132.0  ...                              NaN   
2                        576496.0  ...                              NaN   
3                         72222.0  ...                              NaN   
4                         63653.0  ...                              NaN   
...                           ...  ...                              ...   
3188                     127427.0  ...                              NaN   
3189                      66351.0  ...                              NaN   
3190                      59466.0  ...                              NaN   
3191                      22335.0  ...                              NaN   
3192                      19599.0  ...                              NaN   

      Male population 45-64 raw value  Male population 65+ raw value  \
0                                 NaN                            NaN   
1                                 NaN                            NaN   
2                                 NaN                            NaN   
3                                 NaN                            NaN   
4                                 NaN                            NaN   
...                               ...                            ...   
3188                              NaN                 

<b> Reading data and giving encoding as unicode_escape to remove encoding error<b>

In [4]:
opioid_rate = pd.read_csv("../../../../data/stage_1/2019-Opioid_Rate.csv",encoding='unicode_escape')
opioid_rate

State             County   FIPS  Opiod_Dispensing_Rate
0       AL     Autauga County   1001                  101.3
1       AL     Baldwin County   1003                   67.6
2       AL     Barbour County   1005                   27.2
3       AL        Bibb County   1007                   21.0
4       AL      Blount County   1009                   23.7
...    ...                ...    ...                    ...
3090    WY  Sweetwater County  56037                   70.0
3091    WY       Teton County  56039                   54.6
3092    WY       Uinta County  56041                   59.5
3093    WY    Washakie County  56043                   46.7
3094    WY      Weston County  56045                    4.7

[3095 rows x 4 columns]

<b> Normalize Deaths data by population per 100000 <b>

In [5]:
cause_of_death

Notes                 County  County Code  \
0       NaN     Autauga County, AL         1001   
1       NaN     Autauga County, AL         1001   
2       NaN     Baldwin County, AL         1003   
3       NaN     Baldwin County, AL         1003   
4       NaN     Baldwin County, AL         1003   
...     ...                    ...          ...   
5536    NaN  Sweetwater County, WY        56037   
5537    NaN       Teton County, WY        56039   
5538    NaN       Uinta County, WY        56041   
5539    NaN       Uinta County, WY        56041   
5540    NaN    Washakie County, WY        56043   

                             Drug/Alcohol Induced Cause  \
0     Drug poisonings (overdose) Unintentional (X40-...   
1          Drug poisonings (overdose) Suicide (X60-X64)   
2     Drug poisonings (overdose) Unintentional (X40-...   
3          Drug poisonings (overdose) Suicide (X60-X64)   
4     Drug poisonings (overdose) Undetermined (Y10-Y14)   
...                                                 ...   
5536       Drug poisonings (overdose) Suicide (X60-X64)   
5537  Drug poisonings (overdose) Unintentional (X40-...   
5538  Drug poisonings (overdose) Unintentional (X40-...   
5539  Drug poisonings (overdose) Undetermined (Y10-Y14)   
5540  Drug poisonings (overdose) Unintentional (X40-...   

     Drug/Alcohol Induced Cause Code  Deaths  Population  Crude Rate  
0                                 D1      69     1087149         6.3  
1                                 D2      14     1087149  Unreliable  
2                                 D1     424     3758097        11.3  
3                                 D2      71     3758097         1.9  
4                                 D4      19     3758097  Unreliable  
...                              ...     ...         ...         ...  
5536                              D2      15      873221  Unreliable  
5537                              D1      13      440125  Unreliable  
5538                              D1      51      426347        12.0  
5539                              D4      10      426347  Unreliable  
5540                              D1      14      172114  Unreliable  

[5541 rows x 8 columns]

<b>Normalizing deaths data by population and rounding it off<b>

In [6]:
cause_of_death["Norm_Deaths"] = (cause_of_death["Deaths"]/cause_of_death["Population"]*100000).round()

In [7]:
cause_of_death

Notes                 County  County Code  \
0       NaN     Autauga County, AL         1001   
1       NaN     Autauga County, AL         1001   
2       NaN     Baldwin County, AL         1003   
3       NaN     Baldwin County, AL         1003   
4       NaN     Baldwin County, AL         1003   
...     ...                    ...          ...   
5536    NaN  Sweetwater County, WY        56037   
5537    NaN       Teton County, WY        56039   
5538    NaN       Uinta County, WY        56041   
5539    NaN       Uinta County, WY        56041   
5540    NaN    Washakie County, WY        56043   

                             Drug/Alcohol Induced Cause  \
0     Drug poisonings (overdose) Unintentional (X40-...   
1          Drug poisonings (overdose) Suicide (X60-X64)   
2     Drug poisonings (overdose) Unintentional (X40-...   
3          Drug poisonings (overdose) Suicide (X60-X64)   
4     Drug poisonings (overdose) Undetermined (Y10-Y14)   
...                                                 ...   
5536       Drug poisonings (overdose) Suicide (X60-X64)   
5537  Drug poisonings (overdose) Unintentional (X40-...   
5538  Drug poisonings (overdose) Unintentional (X40-...   
5539  Drug poisonings (overdose) Undetermined (Y10-Y14)   
5540  Drug poisonings (overdose) Unintentional (X40-...   

     Drug/Alcohol Induced Cause Code  Deaths  Population  Crude Rate  \
0                                 D1      69     1087149         6.3   
1                                 D2      14     1087149  Unreliable   
2                                 D1     424     3758097        11.3   
3                                 D2      71     3758097         1.9   
4                                 D4      19     3758097  Unreliable   
...                              ...     ...         ...         ...   
5536                              D2      15      873221  Unreliable   
5537                              D1      13      440125  Unreliable   
5538                              D1      51      426347        12.0   
5539                              D4      10      426347  Unreliable   
5540                              D1      14      172114  Unreliable   

      Norm_Deaths  
0             6.0  
1             1.0  
2            11.0  
3             2.0  
4             1.0  
...           ...  
5536          2.0  
5537          3.0  
5538         12.0  
5539          2.0  
5540          8.0  

[5541 rows x 9 columns]

<b>Creating a deep copy of cause of death dataframe<b>

In [8]:
cause_of_death_new = cause_of_death.copy()

<b>I want to find total deaths per county, and not divided by Drug/Alcohol Induced Cause Code, so I will group the data by County and find the sum of Deaths, Norm_Deaths for each of them. <b>

<b>Grouping the data by County, CountyCode, Population and summing up deaths and Norm_Deaths data<b>
<br><br>
 <b>Keeping as_index = False here. If we donot use it as "False" here then County, County code and population will be taken as index of the dataframe<b>

In [9]:
cause_of_death_new = cause_of_death_new.groupby(["County","County Code","Population"], as_index = False)[["Deaths","Norm_Deaths"]].sum()

In [10]:
cause_of_death_new

County  County Code  Population  Deaths  \
0              Abbeville County, SC        45001      535389      38   
1                 Acadia Parish, LA        22001     1279727     237   
2               Accomack County, VA        51001      726189      73   
3                    Ada County, ID        16001     8083452     988   
4                  Adair County, KY        21001      387950      43   
...                             ...          ...         ...     ...   
2525                Yuba County, CA         6115     1468094     153   
2526  Yukon-Koyukuk Census Area, AK         2290      129587      14   
2527                Yuma County, AZ         4027     3964172     468   
2528              Zapata County, TX        48505      284033      16   
2529              Zavala County, TX        48507      247499      16   

      Norm_Deaths  
0             7.0  
1            19.0  
2            11.0  
3            12.0  
4            11.0  
...           ...  
2525         11.0  
2526         11.0  
2527         12.0  
2528          6.0  
2529          6.0  

[2530 rows x 5 columns]

<b>Merging 2 dataframes. As column names are different, I am mentioning column names to merge on from the frist dataframe and second dataframe<b>

In [11]:
merge1 = pd.merge(cause_of_death_new,health_ranking, left_on = "County Code", right_on = "5-digit FIPS Code")
merge1

County  County Code  Population  Deaths  \
0              Abbeville County, SC        45001      535389      38   
1                 Acadia Parish, LA        22001     1279727     237   
2               Accomack County, VA        51001      726189      73   
3                    Ada County, ID        16001     8083452     988   
4                  Adair County, KY        21001      387950      43   
...                             ...          ...         ...     ...   
2523                Yuba County, CA         6115     1468094     153   
2524  Yukon-Koyukuk Census Area, AK         2290      129587      14   
2525                Yuma County, AZ         4027     3964172     468   
2526              Zapata County, TX        48505      284033      16   
2527              Zavala County, TX        48507      247499      16   

      Norm_Deaths  State FIPS Code  County FIPS Code  5-digit FIPS Code  \
0             7.0               45                 1              45001   
1            19.0               22                 1              22001   
2            11.0               51                 1              51001   
3            12.0               16                 1              16001   
4            11.0               21                 1              21001   
...           ...              ...               ...                ...   
2523         11.0                6               115               6115   
2524         11.0                2               290               2290   
2525         12.0                4                27               4027   
2526          6.0               48               505              48505   
2527          6.0               48               507              48507   

     State Abbreviation                       Name  ...  \
0                    SC           Abbeville County  ...   
1                    LA              Acadia Parish  ...   
2                    VA            Accomack County  ...   
3                    ID                 Ada County  ...   
4                    KY               Adair County  ...   
...                 ...                        ...  ...   
2523                 CA                Yuba County  ...   
2524                 AK  Yukon-Koyukuk Census Area  ...   
2525                 AZ                Yuma County  ...   
2526                 TX              Zapata County  ...   
2527                 TX              Zavala County  ...   

      Male population 18-44 raw value  Male population 45-64 raw value  \
0                                 NaN                              NaN   
1                                 NaN                              NaN   
2                                 NaN                              NaN   
3                                 NaN                              NaN   
4                                 NaN                              NaN   
...                               ...                              ...   
2523                              NaN                              NaN   
2524                              NaN                              NaN   
2525                              NaN                              NaN   
2526                              NaN                              NaN   
2527                              NaN                              NaN   

      Male population 65+ raw value  Total male population raw value  \
0                               NaN                              NaN   
1                               NaN                              NaN   
2                               NaN                              NaN   
3                               NaN                              NaN   
4                               NaN                              NaN   
...                             ...                              ...   
2523                            NaN                              NaN   
2524                            NaN                              NaN   
2525             

<b>Taking only required columns from opioid rate dataframe<b>

In [12]:
opioid_rate = opioid_rate[["State","FIPS","Opiod_Dispensing_Rate"]]
opioid_rate

State   FIPS  Opiod_Dispensing_Rate
0       AL   1001                  101.3
1       AL   1003                   67.6
2       AL   1005                   27.2
3       AL   1007                   21.0
4       AL   1009                   23.7
...    ...    ...                    ...
3090    WY  56037                   70.0
3091    WY  56039                   54.6
3092    WY  56041                   59.5
3093    WY  56043                   46.7
3094    WY  56045                    4.7

[3095 rows x 3 columns]

<b>Finally merging the third dataframe with the merged dataframe of first and second dataframe. Here too, I am mentioning column names as they are different in both dataframes<b>

In [13]:
merge_final = pd.merge(merge1, opioid_rate, left_on = "County Code", right_on = "FIPS")
merge_final

County  County Code  Population  Deaths  \
0              Abbeville County, SC        45001      535389      38   
1                 Acadia Parish, LA        22001     1279727     237   
2               Accomack County, VA        51001      726189      73   
3                    Ada County, ID        16001     8083452     988   
4                  Adair County, KY        21001      387950      43   
...                             ...          ...         ...     ...   
2522                Yuba County, CA         6115     1468094     153   
2523  Yukon-Koyukuk Census Area, AK         2290      129587      14   
2524                Yuma County, AZ         4027     3964172     468   
2525              Zapata County, TX        48505      284033      16   
2526              Zavala County, TX        48507      247499      16   

      Norm_Deaths  State FIPS Code  County FIPS Code  5-digit FIPS Code  \
0             7.0               45                 1              45001   
1            19.0               22                 1              22001   
2            11.0               51                 1              51001   
3            12.0               16                 1              16001   
4            11.0               21                 1              21001   
...           ...              ...               ...                ...   
2522         11.0                6               115               6115   
2523         11.0                2               290               2290   
2524         12.0                4                27               4027   
2525          6.0               48               505              48505   
2526          6.0               48               507              48507   

     State Abbreviation                       Name  ...  \
0                    SC           Abbeville County  ...   
1                    LA              Acadia Parish  ...   
2                    VA            Accomack County  ...   
3                    ID                 Ada County  ...   
4                    KY               Adair County  ...   
...                 ...                        ...  ...   
2522                 CA                Yuba County  ...   
2523                 AK  Yukon-Koyukuk Census Area  ...   
2524                 AZ                Yuma County  ...   
2525                 TX              Zapata County  ...   
2526                 TX              Zavala County  ...   

      Total male population raw value  Female population 0-17 raw value  \
0                                 NaN                               NaN   
1                                 NaN                               NaN   
2                                 NaN                               NaN   
3                                 NaN                               NaN   
4                                 NaN                               NaN   
...                               ...                               ...   
2522                              NaN                               NaN   
2523                              NaN                               NaN   
2524                              NaN                               NaN   
2525                              NaN                               NaN   
2526                              NaN                               NaN   

      Female population 18-44 raw value  Female population 45-64 raw value  \
0                                   NaN                                NaN   
1                                   NaN                                NaN   
2                                   NaN                                NaN   
3                                   NaN                                NaN   
4                                   NaN                                NaN   
...                                 ...                                ...   
2522                                NaN                                NaN   
2523                         

<b>Finally saving the dataframe as csv file inside the data folder of stage_1 with name of superdataset_merge.csv<b>

In [14]:
merge_final.to_csv("../../../../data/stage_1/superdataset_merge.csv", index = False)

 - **M3.2** Identify counties and states (top 10) with the highest opioid mortality rates. (20 pts)
        - Use mean and median for counties within states to compare (for the state level).
        - Describe your intution on why the rates are high in these states and counties. 

#### To calculate opioid mortality rate, we will use Norm_Deaths column. This column gives mortality rate by county.

<b>Checking the columns of merge_final dataset<b>

In [15]:
merge_final.columns

Index(['County', 'County Code', 'Population', 'Deaths', 'Norm_Deaths',
       'State FIPS Code', 'County FIPS Code', '5-digit FIPS Code',
       'State Abbreviation', 'Name',
       ...
       'Total male population raw value', 'Female population 0-17 raw value',
       'Female population 18-44 raw value',
       'Female population 45-64 raw value', 'Female population 65+ raw value',
       'Total female population raw value', 'Population growth raw value',
       'State', 'FIPS', 'Opiod_Dispensing_Rate'],
      dtype='object', length=542)

### Top 10 states:
<b>For state, we will groupby the merged data by state, and summing up all the required columns<b>
<br><br><b> The reason for summing is we want total values of columns for each state (eg: values of deaths, norm_deaths, etc.)<b>

In [16]:
state_data = merge_final.groupby(["State Abbreviation"],as_index = False)[["Norm_Deaths","Opiod_Dispensing_Rate","Population","Primary care physicians raw value","Drug overdose deaths raw value"]].sum()
state_data.head()

State Abbreviation  Norm_Deaths  Opiod_Dispensing_Rate  Population  \
0                 AK        164.0                  329.5    13575363   
1                 AL        545.0                 3825.4    96964024   
2                 AR        573.0                 3938.2    57808448   
3                 AZ        231.0                  502.1   130658024   
4                 CA        834.0                 2071.3   776313263   

   Primary care physicians raw value  Drug overdose deaths raw value  
0                           0.014610                      103.535747  
1                           0.027504                      550.412330  
2                           0.030343                      522.172294  
3                           0.007598                      302.665355  
4                           0.040171                      857.733596

<b>Sorting the groupby data by Norm_Deaths column and showing head(10) as we want to find top 10 states with highest mortality rate.<b>
<br><br><b>Below shows top 10 states with highest mortality rate<b>

In [17]:
top10_states = state_data.sort_values(by = "Norm_Deaths", ascending = False)
top10_states.head(10)

State Abbreviation  Norm_Deaths  Opiod_Dispensing_Rate  Population  \
17                 KY       2209.0                 6348.1    89496904   
42                 TN       1621.0                 5505.7   131143199   
43                 TX       1467.0                 5659.0   512331921   
27                 NC       1384.0                 4641.6   194992265   
45                 VA       1375.0                 6219.4   163866565   
35                 OH       1352.0                 3229.5   241915812   
10                 GA       1323.0                 7161.6   195574543   
15                 IN       1175.0                 4026.2   134586038   
49                 WV       1171.0                 2458.1    38064982   
24                 MO       1116.0                 4084.5   122243559   

    Primary care physicians raw value  Drug overdose deaths raw value  
17                           0.050459                     2851.486869  
42                           0.041835                     1941.596904  
43                           0.073675                      997.407168  
27                           0.052329                     1850.661142  
45                           0.078846                     1576.138962  
35                           0.044694                     2621.528312  
10                           0.064609                     1236.063074  
15                           0.044335                     1800.556458  
49                           0.030135                     1785.084567  
24                           0.043580                     1106.862461

* From above data values we can see that KY is the state with maximum mortality rate (Norm_Deaths). Mortality rate can be calculated by dividing Deaths by population for every data entry(which is done and stored in Norm_Deaths).

<br> The reason of KY having the highest mortality rate is:
* Opioid dispensing rate is high for KY with second lowest population among top 10 states. With more opioid dispensing rate among low population results in high mortality rate. 
* These is one of the reason for all states of being in top 10 in mortality rates.
* KY also has the highest Drug Overdose Deaths which can be seen from the table.

<br>Similary we can figure that out for all other states.

<br> MO is having the lowest mortality rate:
* Opioid dispensing rate is around the average with lower than average population compared to other 9 states. With average dispensing rate on lesser population, it stands at the bottom of the list. 
* Also opioid drug overdose death are also the second lowest which is another reason of MO having least mortality rate among all.

<br> To sum up, these are the factors that I feel impact the mortality rate with many other factors.

<b> M3.2 Use mean and median for counties within states to compare (for the state level).<b>

<b> Extracting state information, as storing it in a list<b>

In [18]:
top_10_states_list = top10_states["State Abbreviation"].head(10).to_list()
top_10_states_list

['KY', 'TN', 'TX', 'NC', 'VA', 'OH', 'GA', 'IN', 'WV', 'MO']

<b>Filtering information corresponding to above state list (10 states) and sorting the values by states<b>

In [19]:
states_subset = merge_final[merge_final["State Abbreviation"].isin(top_10_states_list)].sort_values(by = "State Abbreviation")
states_subset

County  County Code  Population  Deaths  Norm_Deaths  \
2439     White County, GA        13311      545022      76         14.0   
2219  Tattnall County, GA        13267      506960      55         11.0   
2304     Upson County, GA        13293      567492      46          8.0   
846     Gilmer County, GA        13123      579772      78         13.0   
237     Brooks County, GA        13027      335530      11          3.0   
...                   ...          ...         ...     ...          ...   
1961   Ritchie County, WV        54085      214864      17          8.0   
1445    Marion County, WV        54049     1184199     156         13.0   
204      Boone County, WV        54005      511573     242         47.0   
2415     Wayne County, WV        54099      880850     291         33.0   
948    Hancock County, WV        54029      647670     197         31.0   

      State FIPS Code  County FIPS Code  5-digit FIPS Code State Abbreviation  \
2439               13               311              13311                 GA   
2219               13               267              13267                 GA   
2304               13               293              13293                 GA   
846                13               123              13123                 GA   
237                13                27              13027                 GA   
...               ...               ...                ...                ...   
1961               54                85              54085                 WV   
1445               54                49              54049                 WV   
204                54                 5              54005                 WV   
2415               54                99              54099                 WV   
948                54                29              54029                 WV   

                 Name  ...  Total male population raw value  \
2439     White County  ...                              NaN   
2219  Tattnall County  ...                              NaN   
2304     Upson County  ...                              NaN   
846     Gilmer County  ...                              NaN   
237     Brooks County  ...                              NaN   
...               ...  ...                              ...   
1961   Ritchie County  ...                              NaN   
1445    Marion County  ...                              NaN   
204      Boone County  ...                              NaN   
2415     Wayne County  ...                              NaN   
948    Hancock County  ...                              NaN   

      Female population 0-17 raw value  Female population 18-44 raw value  \
2439                               NaN                                NaN   
2219                               NaN                                NaN   
2304                               NaN                                NaN   
846                                NaN                                NaN   
237                                NaN                                NaN   
...                                ...                                ...   
1961                               NaN                                NaN   
1445                               NaN                                NaN   
204                                NaN                                NaN   
2415                               NaN                                NaN   
948                                NaN                                NaN   

      Female population 45-64 raw value  Female population 65+ raw value  \
2439                                NaN                              NaN   
2219                                NaN                              NaN   
2304                                NaN                              NaN   
846                                 NaN                              NaN   
237                                 NaN                              NaN   
...    

<b>Finding the mean and median for each state by and printing it<b>
<br>
<br>
<b> Here I am using a for loop to iterate through each state from the state-list and calculating mean/median of Norm_Deaths<b>
    
<b>I have used format() for mean for it to show upto 2 decimal places<b>

I have used below link for printing till 2 decimal points<br>
https://tutorial.eyehunts.com/python/how-to-display-2-decimal-places-in-python-example-code/#:~:text=Use%20str.,decimal%20places%20in%20the%20console.

In [20]:
for states in top_10_states_list:
    print('Mean of mortality rate of counties in '+states+": " + '{:.2f}'.format(states_subset[states_subset["State Abbreviation"] == states]["Norm_Deaths"].mean()))
    print('Median of mortality rate of county in '+states+": " + str(states_subset[states_subset["State Abbreviation"] == states]["Norm_Deaths"].median()))
    print("")

Mean of mortality rate of counties in KY: 19.04
Median of mortality rate of county in KY: 16.0

Mean of mortality rate of counties in TN: 17.24
Median of mortality rate of county in TN: 16.5

Mean of mortality rate of counties in TX: 8.34
Median of mortality rate of county in TX: 8.0

Mean of mortality rate of counties in NC: 13.98
Median of mortality rate of county in NC: 14.0

Mean of mortality rate of counties in VA: 11.00
Median of mortality rate of county in VA: 9.0

Mean of mortality rate of counties in OH: 15.36
Median of mortality rate of county in OH: 15.0

Mean of mortality rate of counties in GA: 9.73
Median of mortality rate of county in GA: 9.0

Mean of mortality rate of counties in IN: 12.91
Median of mortality rate of county in IN: 12.0

Mean of mortality rate of counties in WV: 22.09
Median of mortality rate of county in WV: 19.0

Mean of mortality rate of counties in MO: 11.05
Median of mortality rate of county in MO: 11.0



<b>Printing county count information for each of ten state using for loop<b>

In [21]:
for i in top_10_states_list:
    print(i+ " has "+ str(len(states_subset[states_subset["State Abbreviation"] == i])) + " counties.")

KY has 116 counties.
TN has 94 counties.
TX has 176 counties.
NC has 99 counties.
VA has 125 counties.
OH has 88 counties.
GA has 136 counties.
IN has 91 counties.
WV has 53 counties.
MO has 101 counties.


- From this we can see that WV has highest mean as well as median. But no. of counties in WV is only 53 because of which it stands at 9th position.
- KY has highest mean and median after WV. It stands at first position because it has total of 116 counties.
- TX has lowest mean and median among all, but still its on 3rd position because it has total of 176 counties which is highest among all.
- Similarly we can compare counties mean and meadian for all other states.

<b>Another approach:<b>
    <br>
<b>For second approach, I try to find top 10 states with higest mean values of mortality rate(Norm_Deaths)<b>

<b>To do that I will use my super dataset and create a deep copy of it<b>

In [39]:
state_mm = merge_final.copy()
state_mm.head()

County  County Code  Population  Deaths  Norm_Deaths  \
0  Abbeville County, SC        45001      535389      38          7.0   
1     Acadia Parish, LA        22001     1279727     237         19.0   
2   Accomack County, VA        51001      726189      73         11.0   
3        Ada County, ID        16001     8083452     988         12.0   
4      Adair County, KY        21001      387950      43         11.0   

   State FIPS Code  County FIPS Code  5-digit FIPS Code State Abbreviation  \
0               45                 1              45001                 SC   
1               22                 1              22001                 LA   
2               51                 1              51001                 VA   
3               16                 1              16001                 ID   
4               21                 1              21001                 KY   

               Name  ...  Total male population raw value  \
0  Abbeville County  ...                              NaN   
1     Acadia Parish  ...                              NaN   
2   Accomack County  ...                              NaN   
3        Ada County  ...                              NaN   
4      Adair County  ...                              NaN   

   Female population 0-17 raw value  Female population 18-44 raw value  \
0                               NaN                                NaN   
1                               NaN                                NaN   
2                               NaN                                NaN   
3                               NaN                                NaN   
4                               NaN                                NaN   

   Female population 45-64 raw value  Female population 65+ raw value  \
0                                NaN                              NaN   
1                                NaN                              NaN   
2                                NaN                              NaN   
3                                NaN                              NaN   
4                                NaN                              NaN   

   Total female population raw value  Population growth raw value  State  \
0                                NaN                          NaN     SC   
1                                NaN                          NaN     LA   
2                                NaN                          NaN     VA   
3                                NaN                          NaN     ID   
4                                NaN                          NaN     KY   

    FIPS  Opiod_Dispensing_Rate  
0  45001                   32.4  
1  22001                   32.2  
2  51001                   19.4  
3  16001                   60.8  
4  21001                   71.3  

[5 rows x 542 columns]

<b>Then I will group the data by States and take mean of certain columns including Norm_Deaths. Finally sorting the values by Norm_Deaths in descending to find top 10 states by Norm_Deaths<b>

In [40]:
state_mm = state_mm.groupby(["State"], as_index = False)[["Norm_Deaths","Opiod_Dispensing_Rate","Population","Primary care physicians raw value","Drug overdose deaths raw value"]].mean()

In [41]:
state_mm.sort_values(by = "Norm_Deaths", ascending = False).head(10)

State  Norm_Deaths  Opiod_Dispensing_Rate    Population  \
49    WV    22.094340              46.379245  7.182072e+05   
7     DC    22.000000              39.900000  1.292383e+07   
32    NM    20.466667              36.196667  1.390853e+06   
17    KY    19.043103              54.725000  7.715250e+05   
8     DE    18.333333              49.700000  6.181811e+06   
20    MD    17.666667              46.350000  5.010535e+06   
42    TN    17.244681              58.571277  1.395140e+06   
19    MA    17.000000              35.592857  9.859132e+06   
39    RI    16.600000              35.080000  4.440928e+06   
6     CT    16.500000              34.775000  9.269304e+06   

    Primary care physicians raw value  Drug overdose deaths raw value  
49                           0.000569                       45.771399  
7                            0.001177                       34.385578  
32                           0.000574                       25.845132  
17                           0.000451                       32.039178  
8                            0.000635                       27.365376  
20                           0.000686                       31.606481  
42                           0.000460                       25.215544  
19                           0.000936                       29.212216  
39                           0.001085                       26.010388  
6                            0.000719                       28.043820

- From above we can see that WV is having the maximum value of mean Norm_Deaths. One reason for it could be it's having very high opioid dispensing rate and also drug overdose death values. These factors contribute to high mortality rate
- Second is DC. It has very high population. It also has high opioid dispensing rate plus high drug overdose deaths. This state has the highest number of rimary care physician. From studies we have seen that primary care physician tend to over prescribe opioids which results in overdose and high mortality rates.
- Similarly, we can decide for all other states.
- CT has the lowest mortality rate because it has the lowest opioid dispensing rate and also very low drug overdose value. It has higher population compared to others. With high population and low overdose and dispensing rates, results in lowest mortality rate.

### Top 10 Counties:
<br>
<b>For county data, as our super data is already at county level and contains only unique county values, we don't need to groupby county or anything. We can directly use it for sorting and taking the max values out.<b>

In [22]:
county_data = merge_final.sort_values(by = "Norm_Deaths", ascending = False)
county_data

County  County Code  Population  Deaths  Norm_Deaths  \
1956  Rio Arriba County, NM        35039      842643     531         63.0   
1486    McDowell County, WV        54047      470658     289         61.0   
2502     Wyoming County, WV        54109      494103     274         56.0   
103      Baltimore city, MD        24510    13091205    6711         51.0   
281       Cabell County, WV        54011     2009558     979         48.0   
...                     ...          ...         ...     ...          ...   
1235    Lagrange County, IN        18087      780851      15          2.0   
1024      Holmes County, OH        39075      884901      19          2.0   
1403     Madison County, ID        16065      735188      18          2.0   
1713   Oktibbeha County, MS        28105      977688      13          1.0   
2159       Starr County, TX        48427     1258845      16          1.0   

      State FIPS Code  County FIPS Code  5-digit FIPS Code State Abbreviation  \
1956               35                39              35039                 NM   
1486               54                47              54047                 WV   
2502               54               109              54109                 WV   
103                24               510              24510                 MD   
281                54                11              54011                 WV   
...               ...               ...                ...                ...   
1235               18                87              18087                 IN   
1024               39                75              39075                 OH   
1403               16                65              16065                 ID   
1713               28               105              28105                 MS   
2159               48               427              48427                 TX   

                   Name  ...  Total male population raw value  \
1956  Rio Arriba County  ...                              NaN   
1486    McDowell County  ...                              NaN   
2502     Wyoming County  ...                              NaN   
103      Baltimore city  ...                              NaN   
281       Cabell County  ...                              NaN   
...                 ...  ...                              ...   
1235    Lagrange County  ...                              NaN   
1024      Holmes County  ...                              NaN   
1403     Madison County  ...                              NaN   
1713   Oktibbeha County  ...                              NaN   
2159       Starr County  ...                              NaN   

      Female population 0-17 raw value  Female population 18-44 raw value  \
1956                               NaN                                NaN   
1486                               NaN                                NaN   
2502                               NaN                                NaN   
103                                NaN                                NaN   
281                                NaN                                NaN   
...                                ...                                ...   
1235                               NaN                                NaN   
1024                               NaN                                NaN   
1403                               NaN                                NaN   
1713                               NaN                                NaN   
2159                               NaN                                NaN   

      Female population 45-64 raw value  Female population 65+ raw value  \
1956                                NaN                              NaN   
1486                                NaN                              NaN   
2502                                NaN                              NaN   
103                                 NaN                              NaN   
281                                 N

<b>Taking only the required columns from county table<b>
<br><b>Below shows top 10 counties with highest mortality rates<b>

In [23]:
county_data[["County","Deaths","Population","Norm_Deaths","Opiod_Dispensing_Rate","Primary care physicians raw value","Drug overdose deaths raw value"]].head(10)

County  Deaths  Population  Norm_Deaths  \
1956  Rio Arriba County, NM     531      842643         63.0   
1486    McDowell County, WV     289      470658         61.0   
2502     Wyoming County, WV     274      494103         56.0   
103      Baltimore city, MD    6711    13091205         51.0   
281       Cabell County, WV     979     2009558         48.0   
204        Boone County, WV     242      511573         47.0   
1364       Logan County, WV     347      752864         46.0   
145         Bell County, KY     269      597689         45.0   
1547       Mingo County, WV     248      554607         45.0   
1521      Mercer County, WV     555     1292174         44.0   

      Opiod_Dispensing_Rate  Primary care physicians raw value  \
1956                   30.8                           0.000375   
1486                   22.7                           0.000470   
2502                   41.5                           0.000276   
103                    71.7                           0.000966   
281                   130.0                           0.001656   
204                    81.5                           0.000394   
1364                   93.7                           0.000415   
145                   155.4                           0.000406   
1547                   69.0                           0.000203   
1521                   65.3                           0.000628   

      Drug overdose deaths raw value  
1956                       75.001685  
1486                       87.059479  
2502                       79.847675  
103                        74.831143  
281                       111.540052  
204                        55.444504  
1364                       70.066711  
145                        55.317893  
1547                       70.207653  
1521                       62.849739

* Rio Arriba County, NM have the highest mortality rate. It has an average population and high opioid dispensing rate as well as drug overdose value. I feel that this is the reason for high mortality rate in this county.
* McDowell County, WV is having second highest mortality rate because it has the least population and high drug overdose date.
* Similary, we can find it out for other counties.
* Mercer County, WV is having the least mortality rate because it has high population compared to others plus it has lower drug overdose rate even with high opioid dispensing rate. This shows that people don't overdose drug which results in lower mortality rate.